### PR TITLE
Recreate inputs

### LINKED ISSUE #14: Text field/Input possible cleanup
I don't think the label and error element shouldn't be generated inside the component based on props, instead they should probably be separate components that you nest (?) inside the TextField element.

It might also make sense to split out TextArea.

A problem that you might run into is the wrapper div around the 3 elements, not sure how to solve that technically.

Example of usage:

```
<label htmlFor="yourid">Some text</label>
<input id="yourid" />
```

should be exchangeable with:

```
<TextFieldLabel htmlFor="yourid">Some text</TextFieldLabel>
<TextField id="yourid" />
```

### PR DESCRIPTION
**Pull request by samrocksc**
#### What's this PR do?

Base recreation of inputs.  Auto Ties them together
#### What are the important parts of the code?

I think the most intersting part is how I tied them together with ID's
#### How should this be tested by the reviewer?

standard npm start
#### Is any other information necessary to understand this?

Not really, context from react.....
#### What are the relevant tickets? (Please add `closes`, `refs`, etc)

Closes #14

### CODE REVIEW

**Comment by samrocksc** [comment]
This is my second submission of TextField.  I used git push -f because I needed to rebase.  I am using context; however, context seems to be geared for situations such as these.

File: example/index.js [hunk 1]
```diff
@@ -52,7 +52,22 @@ const exampleTree = (
...
+        <TextFieldInput/>
+        <TextFieldLabel>Floating Label</TextFieldLabel>
+      </TextFieldContainer>
+      <br/>
+      <TextFieldContainer>
+        <TextFieldInput pattern="-?[0-9]*(\.[0-9]+)?"/>
+        <TextFieldLabel>Error Message</TextFieldLabel>
+        <TextFieldError>Not a number</TextFieldError>
+      </TextFieldContainer>
+      <br/>
```
Comment (phated) [inline example/index.js:70]: Needs an example with TextFieldTextArea

File: src/text-field-container.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+const React = require('react');
+const mdl = require('material-design-lite/material.min');
+const classnames = require('classnames');
+
+const baseClasses = {
+  'mdl-textfield': true,
+  'mdl-js-textfield': true
+};
+
+class TextFieldContainer extends React.Component {
```
Comment (phated) [inline src/text-field-container.js]: Why did you rename this to TextFieldContainer?  I think I like TextField

File: src/text-field-container.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  getChildContext(){
```
Comment (phated) [inline src/text-field-container.js]: don't like context, think we should use `React.cloneElement()` to add props

File: src/text-field-container.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      children,
+      expandable,
+      icon,
+      floating,
+      className
+    } = this.props;
+
+    const classes = classnames(baseClasses, {
+      'mdl-textfield--expandable': expandable,
+      'mdl-button mdl-js-button mdl-button--icon': icon,
```
Comment (phated) [inline src/text-field-container.js]: doesn't this get applied to buttons inside?

File: src/text-field-container.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+TextFieldContainer.propTypes = {
+  expandable: React.PropTypes.bool,
+  icon: React.PropTypes.bool,
+  className: React.PropTypes.string,
+  floating: React.PropTypes.bool
+};
+
+TextFieldContainer.childContextTypes = {
```
Comment (phated) [inline src/text-field-container.js]: Don't like context

File: src/text-field-error.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  render(){
+    const{
+      children,
+      className
+    } = this.props;
+
+    const classes = classnames(baseClasses, className);
+
+    return (
+      <span {...this.props} id={this.context.id} className={classes}>
```
Comment (phated) [inline src/text-field-error.js]: don't like context

File: src/text-field-label.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      icon,
+      children
+    } = this.props;
+
+    const classes = classnames(baseClasses, {
+      'mdl-button mdl-js-button mdl-button--icon': icon
+    }, className);
+
+    return (
+      <label {...this.props} htmlFor={this.context.id} className={classes}>
```
Comment (phated) [inline src/text-field-label.js]: Don't like context

File: src/text-field-label.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      </label>
+    );
+  }
+}
+
+TextFieldLabel.propTypes = {
+  icon: React.PropTypes.bool
+};
+
+TextFieldLabel.contextTypes = {
```
Comment (phated) [inline src/text-field-label.js]: Don't like context

File: example/index.js [hunk 9]
```diff
@@ -1,14 +1,14 @@
 'use strict';
 const React = require('react');
 const ReactDOM = require('react-dom');
-const { Icon, Tooltip, Textfield, Switch, Spinner, Slider, Radio, Progress, IconToggle, Button, Badge, Checkbox } = require('../');
+const { TextFieldContainer, TextFieldInput, TextFieldError, TextFieldLabel, TextFieldTextArea, Icon, Tooltip, Switch, Spinner, Slider, Radio, Progress, IconToggle, Button, Badge, Checkbox } = require('../');
```
Comment (phated) [inline example/index.js]: TextFieldTextArea is destructured wrong (it is properly named TextFieldTextarea everywhere else)

**Comment by phated** [comment]
Need to remove context and use `cloneElement` instead in a bunch of places.  `id` should probably be specified on `propTypes` too.

File: src/text-field-container.js (hunk 2 above)
Comment (samrocksc) [inline src/text-field-container.js]: Because you told me to in the last pull request.

**Comment by samrocksc** [comment]
I didn't realize context was going to be that big of an issue.  I'll work on this later.

File: src/text-field-container.js (hunk 4 above)
Comment (samrocksc) [inline src/text-field-container.js]: There is no need for buttons to have an htmlFor, as they will be toggled inside of a "form component" or use onClick/onChange with a ref value.

Comment (samrocksc) [inline src/text-field-container.js]: By applying it to the parent element it updates the entire element.  I ran tests to make sure on this one.  The code for upgradeElement describes this bettter.

**Comment by samrocksc** [comment]
I updated everything you wanted me to.  There are very obscure references with cloneelement, I think I got the gist of it here.  I also set the ID up to randomly generate at the beginning as a fallback using our pattern we had established with the other toggle items.

I think I got a good pattern with the primary element.

File: src/text-field-textarea.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+  render(){
+    const {
+      className
+    } = this.props;
+
+    const classes = classnames(baseClasses, className);
+
+    return (
+      <textarea {...this.props} type='text' className={classes} />
```
Comment (phated) [inline src/text-field-textarea.js]: textarea doesn't need `type="text"` on it...I don't even think it accepts it

File: src/text-field.js [hunk 11]
```diff
@@ -30,81 +27,48 @@ class Textfield extends React.Component {
...
-      textField = (<textarea className={inputClasses} id={id} />);
-    } else {
-      textField = (<input className={inputClasses} type={type} id={id} />);
-    }
-
-    if (label) {
-      labelField = (<label className={labelClasses} htmlFor={id}>{label}</label>);
-    }
+    const inputChildren = React.Children.map(this.props.children, function(child) {
+      return React.cloneElement(child, {
```
Comment (phated) [inline src/text-field.js]: this is looking good, however, you need to check the type of component each child is.   If the type is TextFieldLabel, you clone with the `htmlFor` property, if it is type TextAreaInput or TextFieldTextarea, you clone with the `id` property.  You also need to check the properties on each child to make sure the `id` or `htmlFor` properties weren't already specified before cloning because we want users to be able to assign their own.

File: src/text-field.js [hunk 12]
```diff
@@ -30,81 +27,48 @@ class Textfield extends React.Component {
...
-      'mdl-textfield--floating-label': floatingLabel,
-      'mdl-textfield--expandable': expandable,
-      'mdl-button': button,
-      'mdl-js-button': button,
-      'mdl-button--icon': icon,
-      'mdl-textfield__expandable-holder': expandableHolder
-    }, className);
-
-    const inputClasses = classnames(inputBaseClasses);
+    let {id} = this.props;
```
Comment (phated) [inline src/text-field.js]: ~~this should be part of the `const` declaration above~~

File: src/text-field.js [hunk 13]
```diff
@@ -30,81 +27,48 @@ class Textfield extends React.Component {
...
-    const labelClasses = classnames(labelBaseClasses);
-
-    let textField;
-    let labelField;
-    let errorField;
-
-    if (!id) {
-      id = '_' + Math.random().toString(36).slice(2);
+    if(!id){
+      id = this._autoId;
```
Comment (phated) [inline src/text-field.js]: use default assignment for the `id` property

File: src/text-field.js [hunk 14]
```diff
@@ -30,81 +27,48 @@ class Textfield extends React.Component {
...
-  floatingLabel: React.PropTypes.bool,
-  error: React.PropTypes.string,
+TextField.propTypes = {
   expandable: React.PropTypes.bool,
-  button: React.PropTypes.bool,
   icon: React.PropTypes.bool,
-  expandableHolder: React.PropTypes.bool
+  className: React.PropTypes.string,
+  floating: React.PropTypes.bool,
+  id: React.PropTypes.string
```
Comment (phated) [inline src/text-field.js]: `id` shouldn't be passed to this element

File: src/text-field.js (hunk 12 above)
Comment (phated) [inline src/text-field.js]: actually, `id` shouldn't be passed to this element

File: src/text-field.js (hunk 13 above)
Comment (phated) [inline src/text-field.js]: This should be removed, the `autoId` is only used for cloning and isn't passed down from props.

**Comment by phated** [comment]
@samrocksc This PR is really shaping up,  just a few more kinks to work out.

File: src/text-field.js (hunk 11 above)
Comment (samrocksc) [inline src/text-field.js]: Due to the fact that that the key element with input and textarea is "name" and "value" keys with html forms.  Also if you choose an onclick or onchange method, you're going to be using a ref method to refer to the input box.  I went ahead and added the check for TextFieldLabel, and kept it strict to that.

**Comment by samrocksc** [comment]
Cool, I got all the changes put in there was a little bit of conflicting feedback on being able to customize the ID on the `<TextArea>` component.  I went ahead and left the ability to assign the ID as a vestigial link in case someone needed access to that variable in the future.

I also made sure that the most pertinent element `TextFieldLabel` to using the id name is properly denoted and passed as `htmlFor`, otherwise it defaults to passing as `id`.

I can change those up if needed; however it seemed like a pretty good compromise and still leaves full customizability to the user.

**Comment by samrocksc** [comment]
this was the last one needed before a version bump i do believe

File: src/text-field.js [hunk 15]
```diff
@@ -30,81 +27,45 @@ class Textfield extends React.Component {
...
+      floating,
+      className,
+      id = this._autoId
     } = this.props;
 
-    let {
-      id
-    } = this.props;
+    const inputChildren = React.Children.map(this.props.children, function(child) {
+      if(child.type.name === 'TextFieldLabel'){
```
Comment (phated) [inline src/text-field.js]: `child.type === TextFieldLabel`  and remember to require TextFieldLabel at the top

File: src/text-field.js [hunk 16]
```diff
@@ -30,81 +27,45 @@ class Textfield extends React.Component {
...
+        return React.cloneElement(child, {htmlFor: id});
+      } else {
+        return React.cloneElement(child, {id: id});
```
Comment (phated) [inline src/text-field.js]: should only add `id` to TextFieldInput and TextFieldTextarea

Comment (phated) [inline src/text-field.js]: should check if `id` already exists on props before cloning and adding a new one

File: src/text-field-textarea.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    const classes = classnames(baseClasses, className);
+
+    return (
+      <textarea {...this.props} className={classes} />
+    );
+  }
+}
+
+TextFieldTextarea.propType = {
+  className: React.PropTypes.string
```
Comment (phated) [inline src/text-field-textarea.js]: this should have an `id` prop type

File: src/text-field.js [hunk 18]
```diff
@@ -30,81 +27,45 @@ class Textfield extends React.Component {
...
-      floatingLabel,
-      error,
+      children,
       expandable,
-      button,
       icon,
-      expandableHolder
+      floating,
+      className,
+      id = this._autoId
```
Comment (phated) [inline src/text-field.js]: this shouldn't be here

File: src/text-field.js [hunk 19]
```diff
@@ -30,81 +27,45 @@ class Textfield extends React.Component {
...
     } = this.props;
 
-    let {
-      id
-    } = this.props;
+    const inputChildren = React.Children.map(this.props.children, function(child) {
```
Comment (phated) [inline src/text-field.js]: i'd probably use an arrow function here and use `this._autoId` inside this function

File: src/text-field-input.js [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    const classes = classnames(baseClasses, className);
+
+    return (
+      <input {...this.props} type='text' className={classes} />
+    );
+  }
+}
+
+TextFieldInput.propType = {
+  className: React.PropTypes.string
```
Comment (phated) [inline src/text-field-input.js]: this should have an `id` prop type

File: src/text-field.js [hunk 21]
```diff
@@ -30,81 +27,45 @@ class Textfield extends React.Component {
...
 
-    const outerClasses = classnames(outerBaseClasses, {
-      'mdl-textfield--floating-label': floatingLabel,
+    const classes = classnames(baseClasses, {
       'mdl-textfield--expandable': expandable,
-      'mdl-button': button,
-      'mdl-js-button': button,
-      'mdl-button--icon': icon,
-      'mdl-textfield__expandable-holder': expandableHolder
+      'mdl-button mdl-js-button mdl-button--icon': icon,
```
Comment (phated) [inline src/text-field.js]: this doesn't apply the to the wrapper, it applies to the label

**Comment by samrocksc** [comment]
changes made

File: src/text-field.js [hunk 22]
```diff
@@ -30,81 +29,58 @@ class Textfield extends React.Component {
...
       expandable,
-      button,
       icon,
-      expandableHolder
-    } = this.props;
-
-    let {
-      id
+      floating,
+      className,
```
Comment (phated) [inline src/text-field.js]: tailing comma

File: src/text-field.js [hunk 23]
```diff
@@ -30,81 +29,58 @@ class Textfield extends React.Component {
...
     } = this.props;
 
-    const outerClasses = classnames(outerBaseClasses, {
-      'mdl-textfield--floating-label': floatingLabel,
+    const inputChildren = React.Children.map(this.props.children, (child) => {
```
Comment (phated) [inline src/text-field.js]: use `children` directly (instead of this.props.children) since you destructured it but didn't use it

File: src/text-field.js [hunk 24]
```diff
@@ -30,81 +29,58 @@ class Textfield extends React.Component {
...
     const {
-      className,
-      type,
-      label,
-      floatingLabel,
-      error,
+      children,
       expandable,
-      button,
       icon,
```
Comment (phated) [inline src/text-field.js]: remove icon, not used

**Comment by samrocksc** [comment]
mop
